### PR TITLE
Use virtual classes for plugins instead of delegates

### PR DESCRIPTION
**Pull request by mikee47**
This PR changes the model slightly to register plugin classes as a group rather than individual events.

* There's no need for a 'register' method as the plugin just overrides the required methods.
* I've added the priority list to HttpResource so there's no need for a separate HttpEventedResource.
* Once a resource instance has been created, plugins are registered directly with it.,

The possible downside is that if events have different priorities they must be implemented in separate plugin instances.

The benefit however is simplification and reduction in RAM usage. I did a check with Host build comparing heap usage for this version and the previous one: 1200 bytes saved - and this RAM would be in use for the lifetime of the HttpServer instance.

I've temporarily modified the application to allocate HttpServer dynamically so this can be checked (also memory leaks by calling shutdown).

### CODE REVIEW

File: Sming/Components/Network/src/Network/Http/Resource/HttpResourcePlugin.h [hunk 1]
```diff
@@ -11,20 +11,64 @@
...
 class HttpResourcePlugin : public LinkedObjectTemplate<HttpResourcePlugin>
 {
 public:
 	using OwnedList = OwnedLinkedObjectListTemplate<HttpResourcePlugin>;
+	class List : public PriorityList<HttpResourcePlugin*>
+	{
+	public:
+		void add(HttpResourcePlugin* plugin)
+		{
+			if(plugin != nullptr) {
```
Comment (slaff) [inline Sming/Components/Network/src/Network/Http/Resource/HttpResourcePlugin.h]: A memory leak is reported here:
```valgrind
==250356==    by 0x13011A: PriorityList<HttpResourcePlugin*>::add(HttpResourcePlugin*, int) (PriorityList.h:40)
==250356==    by 0x12E389: HttpResourcePlugin::List::add(HttpResourcePlugin*) (HttpResourcePlugin.h:31)
==250356==    by 0x131910: void HttpResourcePlugin::List::add<ResourceBasicAuth*>(HttpResourcePlugin*, ResourceBasicAuth*) (HttpResourcePlugin.h:38)
==250356==    by 0x1307D1: HttpResource* HttpResourceTree::set<ResourceBasicAuth*>(String const&, Delegate<void (HttpRequest&, HttpResponse&)> const&, HttpResourcePlugin*, ResourceBasicAuth*) (HttpResourceTree.h:114)
==250356==    by 0x12F602: (anonymous namespace)::startWebServer() (application.cpp:55)
==250356==    by 0x12FA94: (anonymous namespace)::gotIP(IpAddress, IpAddress, IpAddress) (application.cpp:78)
==250356==    by 0x131AA5: std::_Function_handler<void (IpAddress, IpAddress, IpAddress), void (*)(IpAddress, IpAddress, IpAddress)>::_M_invoke(std::_Any_data const&, IpAddress&&, IpAddress&&, IpAddress&&) (std_function.h:300)
==250356==    by 0x1440AC: std::function<void (IpAddress, IpAddress, IpAddress)>::operator()(IpAddress, IpAddress, IpAddress) const (std_function.h:688)
==250356==    by 0x142B27: WifiEventsImpl::stationGotIp(IpAddress, IpAddress, IpAddress) (WifiEventsImpl.h:36)
==250356==    by 0x14349E: StationImpl::statusCallback(netif*) (StationImpl.cpp:143)
==250356==    by 0x142EA2: StationImpl::initialise(netif*)::{lambda(netif*)#1}::operator()(netif*) const::{lambda(unsigned int)#1}::operator()(unsigned int) const (StationImpl.cpp:96)
==250356== 
==250356== LEAK SUMMARY:
==250356==    definitely lost: 16 bytes in 1 blocks
```

File: samples/HttpServer_Plugins/app/application.cpp [hunk 2]
```diff
@@ -50,8 +52,8 @@ void startWebServer()
 	 *
 	 * make sure to replace the IP address with the IP address of your HttpServer
 	*/
-	server.paths.set("/ip-n-auth", echoContentBody,
-					 new ResourceIpAuth(IpAddress("192.168.13.0"), IpAddress("255.255.255.0")), pluginBasicAuth);
+	server->paths.set("/ip-n-auth", echoContentBody,
```
Comment (slaff) [inline samples/HttpServer_Plugins/app/application.cpp:55]: Here the Basic Auth plugins seems to be ignored. You can reproduce it by calling:
```
curl -vvv http://username:@192.168.13.10/ip-n-auth -d "1234" -H "Content-Encoding: test"
```

The request should be allowed only when the username and password match. In comparison this curl request should work:
```
curl -vvv http://username:password@192.168.13.10/ip-n-auth -d "1234" -H "Content-Encoding: test"
```

Comment (mikee47) [reply samples/HttpServer_Plugins/app/application.cpp:55]: Bug in `LinkedObject::insertAfter()`, fixed.

File: Sming/Components/Network/src/Network/Http/Resource/HttpResourcePlugin.h (hunk 1 above)
Comment (mikee47) [reply Sming/Components/Network/src/Network/Http/Resource/HttpResourcePlugin.h]: Not sure if this is fixed or not... steps to reproduce?

**Comment by mikee47** [comment]
@slaff Could we simplify this further, perhaps? These plugins seem to fit into either a `pre-filter` or `post-filter` category, depending on whether they're called before or after the resource itself. If that holds true then we could just implement one of two base classes (`HttpPreFilter` and `HttpPostFilter`). The `getPriority()` method would return 'pre' or 'post' accordingly (no need to store the 'priority' value either).

**Comment by slaff** [comment]
> Could we simplify this further, perhaps? These plugins seem to fit into either ...

Ok, give it a try.

File: Sming/Components/Network/src/Network/Http/Resource/HttpResourcePlugin.h (hunk 1 above)
Comment (slaff) [reply Sming/Components/Network/src/Network/Http/Resource/HttpResourcePlugin.h]: > Not sure if this is fixed or not... steps to reproduce?

Seems to be working. But here is how you can test it yourself:
```
cd $SMING_HOME/../samples/HttpServer_Plugins
make SMING_ARCH=Host ENABLE_GDB=1 DEBUG_VERBOSE_LEVEL=3 ENABLE_MALLOC_COUNT=0
make valgrind
```

Now run in a separate terminal window:
```
curl -vvv http://username:@192.168.13.10/ip-n-auth -d "1234" -H "Content-Encoding: test"
curl -vvv http://username:password@192.168.13.10/ip-n-auth -d "1234" -H "Content-Encoding: test"
curl -vvv http://username:password@192.168.13.10/test -d "1234" -H "Content-Encoding: test"
```

And then stop the application. Valgrind will give you a memory leak report.

**Pull request merged**